### PR TITLE
feat: build core erc4626-resolver

### PR DESCRIPTION
**Pull request by letteldream**
Build ERC4626 resolver
- get Vault details
- get user balance of underlying token & Vault deposited.
- get user token allowance of Vault

### CODE REVIEW

File: contracts/protocols/mainnet/erc4626/main.sol [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            isToken = false;
+        }
+
+        _vaultData = VaultData(
+            isToken,
+            vaultToken.name(),
+            vaultToken.symbol(),
+            vaultToken.decimals(),
+            vaultToken.asset(),
+            vaultToken.totalAssets()
```
Comment (KaymasJain) [inline contracts/protocols/mainnet/erc4626/main.sol]: Hey! So as resolvers are used for read calls. You can return all the data. For example, convert to assets or convert to shares as well. As this is also some data we will show users on UI. Just try to return as much things as possible as the gas is not an issue here.

Comment (KaymasJain) [reply contracts/protocols/mainnet/erc4626/main.sol]: For convert to assets & convert to shares. You can use 1 token for example 10**decimals.

File: contracts/protocols/mainnet/erc4626/main.sol [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        );
+
+        return _vaultData;
+    }
+
+    function getPosition(address owner, address vaultAddress) public view returns (uint256, uint256) {
```
Comment (KaymasJain) [inline contracts/protocols/mainnet/erc4626/main.sol]: Similar to how we do on ERC20 resolver. Also add a function as `getPositions(address owner, address[] vaultAddresses)` to fetch mutiple vault's data at once.

File: contracts/protocols/mainnet/erc4626/main.sol [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    struct VaultData {
+        bool isToken;
+        string name;
+        string symbol;
+        uint256 decimals;
+        address asset;
+        uint256 totalAssets;
+    }
+
+    function getVaultDetails(address vaultAddress) public view returns (VaultData memory) {
```
Comment (KaymasJain) [inline contracts/protocols/mainnet/erc4626/main.sol]: Also, add one more function as to fetch mutliple vault's detail at once.
`getVaultsDetails(address[] vaultAddresses)`

File: contracts/protocols/mainnet/erc4626/main.sol (hunk 1 above)
Comment (KaymasJain) [reply contracts/protocols/mainnet/erc4626/main.sol]: Also, the data of `convert to assets` & `convert to shares` also helps on the UI level when user wants to deposit/mint & withdraw/redeem to set the minimum or maximum limit after particular action.

File: contracts/protocols/mainnet/erc4626/main.sol [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        }
+
+        try vaultToken.name() {} catch {
+            isToken = false;
+        }
+
+        try vaultToken.decimals() {} catch {
+            isToken = false;
+        }
+
```
Comment (thrilok209) [inline contracts/protocols/mainnet/erc4626/main.sol:40]: if `isToken` is false, return here. Don't proceed further.

Comment (thrilok209) [reply contracts/protocols/mainnet/erc4626/main.sol:40]: Also, validate if it has exchangePrice, asset, etc as well.

File: contracts/protocols/mainnet/erc4626/main.sol [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        uint256 _tokenAllowance;
+
+        VaultInterface vaultToken = VaultInterface(vaultAddress);
+        _tokenAllowance = vaultToken.allowance(owner, vaultAddress);
+        return _tokenAllowance;
+    }
+}
+
+contract InstaERC4626Resolver is Resolver {
+    string public constant name = "ERC4626-Resolver-v1.1";
```
Comment (thrilok209) [inline contracts/protocols/mainnet/erc4626/main.sol]:
```suggestion
    string public constant name = "ERC4626-Resolver-v1";
```

File: contracts/protocols/mainnet/erc4626/main.sol (hunk 4 above)
Comment (letteldream) [reply contracts/protocols/mainnet/erc4626/main.sol:31]: I think we can only validate vaulToken.asset() like this:
            
try vaultToken.asset() {} catch {
    isToken = false;
    continue;
}

And what about the validating exchangePrice?

**Comment by letteldream** [comment]
ERC4626 Resolver contract address:
https://etherscan.io/address/0xd35098360158476d0b05542dcd5829907e1f75ac

**Review by shriyatyagii — APPROVED** [review]

**Pull request merged**